### PR TITLE
trie: encode and decode trie for DB, save and load to/from DB

### LINKED ISSUE #324: persist state trie between restarts
- [ ] when the node shuts down, save the state trie to the DB  (SCALE entire trie??)
- [ ] when it starts up again, load it from the DB

### PR DESCRIPTION
**Pull request by noot**
## Changes
- implement Decode functions for trie nodes (note that due to the recursive nature of branch encoding, it contains the hashes of its children, which can't be reconstructed. instead, the Decode function stubs where it knows its child should be with an empty node) either way, this function is only used in DecodeFromDB where the trie is recursively reconstructed
- implement EncodeForDB and DecodeFromDB. Encode recursively encodes each node in the trie and SCALEs it and appends it to a byte array. Decode performs the opposite, it reads each node one at a time from the encoding and reconstructs the trie
- implement StoreInDB and LoadFromDB. Store saves the trie in the DB at its root hash and Load loads a trie given a root hash. (later on, when we need to restore a certain state, we can do it from knowing the state hash)

## Tests:
```
go test ./trie/...
```

### Issues:
closes #324

### CODE REVIEW

File: trie/database.go [hunk 1]
```diff
@@ -30,64 +34,124 @@ type StateDB struct {
...
+	if err != nil {
+		return err
+	}
+
+	encroot, err := t.EncodeRoot()
+	if err != nil {
+		return err
+	}
+
+	return t.db.Db.Put(encroot, enc)
```
Comment (ansermino) [inline trie/database.go]: I think we also need to store the encoded root in the db. Probably something like whats here: https://github.com/ChainSafe/gossamer/blob/development/core/rawdb/schema.go

File: trie/print.go [hunk 2]
```diff
@@ -20,6 +20,26 @@ import (
 	"fmt"
 )
 
+// String returns the trie stringified through pre-order traversal
+func (t *Trie) String() string {
```
Comment (ansermino) [inline trie/print.go]: Conflicts with `print.go`, please merge. Would recommend using gotree for cleaner output

File: trie/trie.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func (t *Trie) Db() *StateDB {
 }
 
 // Encode returns the encoded root of the trie
-func (t *Trie) Encode() ([]byte, error) {
+func (t *Trie) EncodeRoot() ([]byte, error) {
```
Comment (ansermino) [inline trie/trie.go]: Update comment

File: trie/database.go (hunk 1 above)
Comment (noot) [reply trie/database.go]: I'm not sure why?

File: trie/print.go (hunk 2 above)
Comment (noot) [reply trie/print.go]: I merged with development and I don't see any conflicts?

File: trie/trie.go (hunk 3 above)
Comment (noot) [reply trie/trie.go]: done

File: trie/print.go (hunk 2 above)
Comment (ansermino) [reply trie/print.go]: Merge the functions, they do basically the same thing

File: trie/database.go (hunk 1 above)
Comment (ansermino) [reply trie/database.go]: It needs to implemented for this to ever work

Comment (noot) [reply trie/database.go]: I'm not sure I see why

Comment (ansermino) [reply trie/database.go]: How will you know what the root hash is? It needs to be stored at a fixed location in the db

Comment (noot) [reply trie/database.go]: do you mean the latest root hash? or every root hash?

I was going to add some functionality to get the latest root hash in my next PR by storing it at a fixed key, since this PR isn't dealing with the actual state

Comment (noot) [reply trie/database.go]: I have added functionality to store/load the hash from a fixed key

File: trie/print.go (hunk 2 above)
Comment (noot) [reply trie/print.go]: okay done

File: trie/node.go [hunk 4]
```diff
@@ -185,6 +187,127 @@ func (l *leaf) Encode() ([]byte, error) {
...
+// least significant six bits of first byte: if len(key) > 62, 0x3f, otherwise len(key)
+// where Extra partial key length is included if len(key) > 63: consists of the remaining key length
+// note that partial key length is the length of the pk in nibbles
+// Partial Key is the branch's key
+// Value is:
+// Children Bitmap | SCALE Branch Node Value | Hash(Enc(Child[i_1])) | Hash(Enc(Child[i_2])) | ... | Hash(Enc(Child[i_n]))
+// Note that since the encoded branch stores the hash of the children nodes, we aren't able to reconstruct the child
+// nodes from the encoding. This function instead stubs where the children are known to be with an empty leaf.
+func (b *branch) Decode(r io.Reader, header byte) (err error) {
+	if header == 0 {
```
Comment (amerameen) [inline trie/node.go]: in which case would we have header == 0?

File: trie/node.go [hunk 5]
```diff
@@ -185,6 +187,127 @@ func (l *leaf) Encode() ([]byte, error) {
...
+			return err
+		}
+	}
+
+	nodeType := header >> 6
+	if nodeType != 2 && nodeType != 3 {
+		return fmt.Errorf("cannot decode node to branch")
+	}
+
+	keyLen := header & 0x3f
```
Comment (amerameen) [inline trie/node.go]: Can you explain what's going on here? Can't seem to understand why we would always add 0x3f to the header.

Comment (noot) [reply trie/node.go]: its a mask, it sets the top 2 bits to 0, since the top 2 bits are the node header and the last 6 bits are the length
0x3f = 0b00111111

File: trie/node.go (hunk 4 above)
Comment (noot) [reply trie/node.go]: if the header hasn't been read yet, if we pass in a value for header it means we already read it from the reader, so we don't want to read another byte for it

File: trie/node.go [hunk 6]
```diff
@@ -111,11 +114,10 @@ func Encode(n node) ([]byte, error) {
 // where NodeHeader is a byte:
 // most significant two bits of first byte: 10 if branch w/o value, 11 if branch w/ value
 // least significant six bits of first byte: if len(key) > 62, 0x3f, otherwise len(key)
-// where Extra partial key length is included if len(key) > 63:
-// consists of the remaining key length
+// where Extra partial key length is included if len(key) > 63: consists of the remaining key length
 // Partial Key is the branch's key
 // Value is:
-// Children Bitmap | Enc(Child[i_1]) | Enc(Child[i_2]) | ... | Enc(Child[i_n]) | SCALE Branch Node Value
+// Children Bitmap | SCALE Branch Node Value | Hash(Enc(Child[i_1])) | Hash(Enc(Child[i_2])) | ... | Hash(Enc(Child[i_n]))
```
[the remark this answers is not in the document]
Comment (noot) [reply trie/node.go]: it's wrong and should be updated, when we were going over the trie integration tests we found out the value comes before

File: trie/database.go [hunk 7]
```diff
@@ -30,64 +34,124 @@ type StateDB struct {
...
-// Commit writes the contents of the db's batch writer to the db
-func (t *Trie) Commit() error {
-	return t.db.Batch.Write()
+func decode(r io.Reader, prev node) error {
+	sd := &scale.Decoder{Reader: r}
+
+	if b, ok := prev.(*branch); ok {
+		for i, child := range b.children {
+			if child != nil {
+				// there's supposed to be a child here, decode the next node and place it
```
[the remark this answers is not in the document]
Comment (noot) [reply trie/database.go]: when we decode a branch node, we only know if a child is supposed to exist at a certain index (due to the bitmap). we also have the hashes of the children, but we can't reconstruct the children from that. so instead, I just put an empty leaf node where the child should be, so when we reconstruct it in this function, we can see that it's non-nil and we should decode the next node from the reader and place it here

File: trie/database.go [hunk 8]
```diff
@@ -30,64 +34,124 @@ type StateDB struct {
...
+	_, err = n.Write(scroot.([]byte))
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	t.db.Lock.Lock()
-	err = t.db.Batch.Put(hash[:], encRoot)
-	t.db.Lock.Unlock()
+	t.root, err = Decode(n)
```
[the remark this answers is not in the document]
Comment (noot) [reply trie/database.go]: I added a comment to Decode to make it more clear, I'd like to leave it as it's named now but happy to discuss

File: trie/database.go [hunk 9]
```diff
@@ -30,64 +34,124 @@ type StateDB struct {
...
 
-// writeNodeToDB returns true if node is written to db batch writer, false otherwise
-// if node is clean, it will not attempt to be written to the db
-// otherwise if it's dirty, try to write it to db
-func (t *Trie) writeNodeToDB(n node) (bool, error) {
-	if !n.isDirty() {
-		return false, nil
+// Decode decodes a trie from the DB and sets the receiver to it
+// The encoded trie must have been encoded with EncodeForDB
+func (t *Trie) Decode(enc []byte) error {
```
[the remark this answers is not in the document]
Comment (noot) [reply trie/database.go]: since the Trie struct also has a DB behind it, so this allows someone to make a trie with a DB and then get the encoding from the DB, and decode it back into the trie

File: trie/database.go [hunk 10]
```diff
@@ -30,64 +34,124 @@ type StateDB struct {
...
 }
 
-// writeNodeToDB returns true if node is written to db batch writer, false otherwise
-// if node is clean, it will not attempt to be written to the db
-// otherwise if it's dirty, try to write it to db
-func (t *Trie) writeNodeToDB(n node) (bool, error) {
-	if !n.isDirty() {
-		return false, nil
+// Decode decodes a trie from the DB and sets the receiver to it
+// The encoded trie must have been encoded with EncodeForDB
```
[the remark this answers is not in the document]
Comment (noot) [reply trie/database.go]: yes I'll change it

File: trie/database.go [hunk 11]
```diff
@@ -30,64 +34,124 @@ type StateDB struct {
...
+	}
+
+	scnenc, err := scale.Encode(nenc)
+	if err != nil {
+		return nil, err
 	}
 
+	enc = append(enc, scnenc...)
+
 	switch n := n.(type) {
```
[the remark this answers is not in the document]
Comment (noot) [reply trie/database.go]: i could also do `if ok, b := n.(*branch); ok`, I feel both are the same verbosity but either would work really

**Comment by ansermino** [comment]
Please fix `TestExt_storage_root`, looking into the other failure...

**Comment by noot** [comment]
@ansermino  fixed

**Comment by noot** [comment]
core tests are still failing on CI but pass for me locally :/

**Comment by ansermino** [comment]
@noot failed locally for me once, but can't reproduce. Re-ran CI, looks good :+1:

File: trie/database.go [hunk 12]
```diff
@@ -17,11 +17,18 @@
...
+	"fmt"
+	"io"
 	"sync"
 
+	scale "github.com/ChainSafe/gossamer/codec"
+	"github.com/ChainSafe/gossamer/common"
 	"github.com/ChainSafe/gossamer/polkadb"
 )
 
+var LATEST_HASH_KEY = []byte("latest_hash")
```
Comment (ansermino) [inline trie/database.go]: Please use go conventions (`LatestHashKey`)

File: trie/database.go [hunk 13]
```diff
@@ -61,33 +133,46 @@ func (t *Trie) writeToDB(n node) error {
...
+		return err
 	}
 
-	t.db.Lock.Lock()
-	err = t.db.Batch.Put(hash[:], encRoot)
-	t.db.Lock.Unlock()
+	return t.Decode(enctrie)
+}
+
+func (t *Trie) StoreHash() error {
```
Comment (ansermino) [inline trie/database.go]: Please document

File: trie/print.go [hunk 14]
```diff
@@ -20,67 +20,60 @@ import (
...
-			fmt.Printf("%x", b)
-		}
-	}
-	fmt.Println("]")
+// Print prints the trie through pre-order traversal
+func (t *Trie) Print() {
+	fmt.Println(t.String())
+}
+
+func (t *Trie) PrintEncoding() {
```
Comment (ansermino) [inline trie/print.go]: Comment

File: trie/database.go [hunk 15]
```diff
@@ -61,33 +133,46 @@ func (t *Trie) writeToDB(n node) error {
...
 
-	n.setDirty(false)
-	return true, err
+	return t.db.Db.Put(LATEST_HASH_KEY, hash[:])
 }
 
-// Commit writes the contents of the db's batch writer to the db
-func (t *Trie) Commit() error {
-	return t.db.Batch.Write()
+func (t *Trie) LoadHash() (common.Hash, error) {
```
Comment (ansermino) [inline trie/database.go]: Please comment. I think `LoadLatestHash` might be less ambiguous

File: trie/database.go [hunk 16]
```diff
@@ -17,11 +17,18 @@
...
 
+	scale "github.com/ChainSafe/gossamer/codec"
+	"github.com/ChainSafe/gossamer/common"
 	"github.com/ChainSafe/gossamer/polkadb"
 )
 
+var LATEST_HASH_KEY = []byte("latest_hash")
+
 // StateDB is a wrapper around a polkadb
 type StateDB struct {
```
Comment (ansermino) [inline trie/database.go]: This should get updated and used. It should probably be called `Database`. I think this should provide `LoadRootHash()`,  `SaveRootHash()` and anything else that is not direclty relevant for the Trie to function. We can probably also remove most of its fields.

The reason we want to make sure to have an extra layer between the trie and DB is so we can implement caching further down the road.

File: trie/database.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (amerameen) [reply trie/database.go]: that makes sense! thanks for explaining

File: trie/database.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (amerameen) [reply trie/database.go]: Yeah, I don't know really what's best; I just found it hard to follow without going at looking at the source code for each of these functions and figuring out what they were doing.

File: trie/database.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (amerameen) [reply trie/database.go]: Ahhh, that make sense to me now! Perhaps the comment could be clearer?

File: trie/database.go [hunk 17]
```diff
@@ -138,20 +141,38 @@ func (t *Trie) StoreInDB() error {
...
 	}
 
-	encroot, err := t.EncodeRoot()
+	roothash, err := t.Hash()
 	if err != nil {
 		return err
 	}
 
-	return t.db.Db.Put(encroot, enc)
+	return t.db.Db.Put(roothash[:], enc)
```
Comment (amerameen) [inline trie/database.go]: This makes a lot more sense than the encroot! Thanks for making that change.

File: trie/database.go (hunk 15 above)
Comment (noot) [reply trie/database.go]: done

File: trie/print.go (hunk 14 above)
Comment (noot) [reply trie/print.go]: done

File: trie/database.go (hunk 13 above)
Comment (noot) [reply trie/database.go]: done

File: trie/database.go (hunk 12 above)
Comment (noot) [reply trie/database.go]: done

File: trie/database.go (hunk 16 above)
Comment (noot) [reply trie/database.go]: okay, the issue with LoadRootHash from the DB is we would have to pass in the root hash, we can't get it directly from the trie. also the other services are interacting with the trie only, I think it's a bit nicer to just have it accessible from the trie rather than doing trie.Db.stuff

File: trie/database.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (noot) [reply trie/database.go]: I'll update it

**Comment by noot** [comment]
@ansermino lemme know how to turn the comments into a package-level comment, or I could just move the fat comment to the node interface?

File: trie/database.go (hunk 16 above)
Comment (noot) [reply trie/database.go]: I have made the requested changes upon discussion, let me know what you think

**Comment by noot** [comment]
@ansermino I've turned the node.go comments into a package-level comment, let me know what you think
